File: src/main/java/org/example/shapecarvejavafx/ShapeCarver.java
package org.example.shapecarvejavafx;

import java.io.*;
import java.util.*;
import javax.script.*;

import org.openjdk.engine.python.*;
import org.openjdk.engine.python.AbstractPythonScriptEngine.PyExecMode;

public class ShapeCarver {
    public static String toZyxString(Object o) {
        var res = o.toString();
        return res.replace(" ", "");
    }

    public static void main(String[] args) throws Exception {
        if (args.length != 1) {
            System.err.println("Usage: ShapeCarver <test_case_index>");
            System.exit(1);
        }

        var m = new ScriptEngineManager();
        var e = (PythonScriptEngine) m.getEngineByName("python");
        e.setExecMode(PyExecMode.SINGLE);
        e.eval("import numpy as np");
        e.setExecMode(PyExecMode.EVAL);

        var testCaseIndex = Integer.parseInt(args[0]);
        var c = new ShapeCarver();
        var process = new ProcessBuilder("./venv/bin/python3", "test_case_to_java_obj.py", args[0]).start();
        try (InputStream stdout = process.getInputStream();
             InputStream stderr = process.getErrorStream()) {

            // We transfer the bytes of the diff directly to System.out
            stdout.transferTo(System.out);
            // In case of execution issues, transfer error streams to System.err
            stderr.transferTo(System.err);
        }
        int exitCode = process.waitFor();
        if (exitCode != 0) {
            System.exit(exitCode);
        }
        var s = new ObjectInputStream(new FileInputStream("roundtrip.ser"));
        var views = (int[][][]) s.readObject();
        var viewsAsList = new ArrayList<List<Integer>>();
        for (int[][] view : views) {
            var viewAsList = new ArrayList<Integer>();
            for (int[] row : view) {
                for (int color : row) {
                    viewAsList.add(color);
                }
            }
            viewsAsList.add(viewAsList);
        }
        // assuming views:
        // 0 (z, y)
        // 1 (x, z)
        // 2 (y, x)
        c.volumeDims = e.newPyTuple(e.fromJava(views[2].length), e.fromJava(views[0][0].length), e.fromJava(views[0].length));
        var intDType = (PyObject) e.eval("int");
        var npDotEmpty = (PyObject) e.eval("np.empty");
        c.volume = npDotEmpty.call(c.volumeDims, intDType);

        var maskColor = e.fromJava(0);
        var npViews = viewsToNumPy(e, viewsAsList);
        var npZeros = (PyObject) e.eval("np.zeros");
        var boolDType = (PyObject) e.eval("bool");
        var skip = npZeros.call(e.newPyTuple(e.fromJava(3), e.fromJava(2)), boolDType);
        e.put("res", c.carve(e, npViews, maskColor, skip));
        var printWriter = new PrintWriter("volume.txt");
        printWriter.println(toZyxString(e.eval("res.tolist()")));
        printWriter.close();

        try {
            // Define the processes to run in the pipeline
            List<ProcessBuilder> builders = Arrays.asList(
                    new ProcessBuilder("jq", "-c", ".[" + testCaseIndex + "].want", "test_cases.json")
                    , new ProcessBuilder("gsed", "s/-1\\>/16711935/g")
                    , new ProcessBuilder("git", "diff", "volume.txt", "/dev/stdin")
            );

            // startPipeline hooks the output of 'jq' to the input of 'git' at the OS level
            List<Process> processes = ProcessBuilder.startPipeline(builders);

            Process jqProcess = processes.get(0);
            Process gsedProcess = processes.get(1);
            Process gitProcess = processes.get(2);

            // Read the output of the final process (git diff) and write it to our stdout
            try (InputStream stdout = gitProcess.getInputStream();
                 InputStream stderr = gitProcess.getErrorStream()) {

                // We transfer the bytes of the diff directly to System.out
                stdout.transferTo(System.out);
                // In case of execution issues, transfer error streams to System.err
                stderr.transferTo(System.err);

                // Also print errors from jq if any occurred
                try (InputStream jqStderr = jqProcess.getErrorStream()) {
                    jqStderr.transferTo(System.err);
                }

                try (InputStream gsedStderr = gsedProcess.getErrorStream()) {
                    gsedStderr.transferTo(System.err);
                }
            }

            int jqExitCode = jqProcess.waitFor();
            int gsedExitCode = gsedProcess.waitFor();
            int gitExitCode = gitProcess.waitFor();

            if (jqExitCode != 0) {
                System.err.printf("jq process failed with exit code: %d%n", jqExitCode);
                System.exit(jqExitCode);
            }

            if (gsedExitCode != 0) {
                System.err.printf("gsed process failed with exit code: %d%n", gsedExitCode);
                System.exit(gsedExitCode);
            }

            if (gitExitCode != 0) {
                System.exit(1);
            }

        } catch (IOException ex) {
            System.err.println("Pipeline execution error (missing binary or invalid file): " + ex.getMessage());
            System.exit(1);
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
            System.err.println("Execution was interrupted.");
            System.exit(1);
        }
    }

    PyTuple volumeDims; /* cuboid shape */
    // Type: MArray{Tuple{S, S, S}, Int}
    PyObject volume;

    // note that JavaFX uses a y-down coordinate system, so the views are left, right, top, bottom, front, back
    public PyObject carve(
            PythonScriptEngine e,
            // Type: SArray{Tuple{3, 2, S, S}, Int}
            PyObject views /* 2d images {x,y,z}-{front,back} */,
            final PyObject maskColor,
            // Type: SArray{Tuple{3, 2}, Bool}
            PyObject skip /* views to skip */) throws ScriptException, NoSuchMethodException {
        Objects.requireNonNull(views);
        Objects.requireNonNull(skip);


        //Initialize volume. This is necessary.
        e.invokeMethod(volume, "fill", e.fromJava(-1));

        var npDotEmpty = (PyObject) e.eval("np.empty");
        var intDType = (PyObject) e.eval("int");
        // Type: MArray{Tuple{3, 2, S, S}, Int}
        var depths = npDotEmpty.call(views.getAttribute("shape"), intDType);
        // Type: MVector{3, Int}
        var cursor = npDotEmpty.call(e.fromJava(3), intDType); // (z, y, x)
        //Initialize depth fields
        var dIt = e.invokeMethod(e.invokeFunction("range", e.fromJava(3)), "__iter__");
        while (true) {
            try {
                var projAxis = (PyObject) e.invokeMethod(dIt, "__next__");
                var u = (PyObject) e.invokeMethod((PyObject) e.invokeMethod(projAxis, "__add__", e.fromJava(1)), "__mod__", e.fromJava(3)); // other axis 0
                var v = (PyObject) e.invokeMethod((PyObject) e.invokeMethod(projAxis, "__add__", e.fromJava(2)), "__mod__", e.fromJava(3)); // other axis 1
                var sideIdxIt = e.invokeMethod(e.invokeFunction("range", e.fromJava(2)), "__iter__");
                while (true) {
                    try {
                        var sideIdx = (PyObject) e.invokeMethod(sideIdxIt, "__next__");
                        var idxTuple = e.newPyTuple(projAxis, sideIdx);
                        var view = views.getItem(idxTuple);
                        var isMinSide = (PyObject) e.invokeMethod(sideIdx, "__eq__", e.fromJava(0));
                        var isMaxSide = (PyObject) e.invokeMethod(sideIdx, "__eq__", e.fromJava(1));
                        var nearBoundary = (PyObject) (isMaxSide.isTrue() ? e.invokeMethod(volumeDims.getItem(projAxis), "__sub__", e.fromJava(1)) : e.fromJava(0));
                        var farBoundary = (PyObject) (isMinSide.isTrue() ? e.invokeMethod(volumeDims.getItem(projAxis), "__sub__", e.fromJava(1)) : e.fromJava(0));
                        // TODO: Switch u and v here to match cursor semantics below
                        var uIdxIt = e.invokeMethod(e.invokeFunction("range", volumeDims.getItem(u)), "__iter__");
                        while (true) {
                            try {
                                var uIdx = (PyObject) e.invokeMethod(uIdxIt, "__next__");
                                var vIdxIt = e.invokeMethod(e.invokeFunction("range", volumeDims.getItem(v)), "__iter__");
                                while (true) {
                                    try {
                                        var vIdx = (PyObject) e.invokeMethod(vIdxIt, "__next__");
                                        var shouldSkip = skip.getItem(idxTuple);
                                        var viewColor = view.getItem(e.newPyTuple(uIdx, vIdx));
                                        var isMaskColor = (PyObject) e.invokeMethod(viewColor, "__eq__", maskColor);
                                        depths.setItem(e.newPyTuple(projAxis, sideIdx, uIdx, vIdx),
                                                shouldSkip.isFalse() && isMaskColor.isTrue() ? farBoundary : nearBoundary);
                                    } catch (Exception ex) {
                                        break;
                                    }
                                }
                            } catch (Exception ex) {
                                break;
                            }
                        }
                    } catch (Exception ex) {
                        break;
                    }
                }

                //Clear out volume where ray goes through entirely
                var cursorVIt = e.invokeMethod(e.invokeFunction("range", volumeDims.getItem(v)), "__iter__");
                while (true) {
                    try {
                        var cursorV = (PyObject) e.invokeMethod(cursorVIt, "__next__");
                        cursor.setItem(v, cursorV);
                        var cursorUIt = e.invokeMethod(e.invokeFunction("range", volumeDims.getItem(u)), "__iter__");
                        while (true) {
                            try {
                                var cursorU = (PyObject) e.invokeMethod(cursorUIt, "__next__");
                                cursor.setItem(u, cursorU);
                                var startDepth = depths.getItem(e.newPyTuple(projAxis, e.fromJava(1), cursor.getItem(v), cursor.getItem(u)));
                                var cursorDIt = e.invokeMethod(e.invokeFunction("range", startDepth, e.invokeMethod(depths.getItem(e.newPyTuple(projAxis, e.fromJava(0), cursor.getItem(v), cursor.getItem(u))), "__add__", e.fromJava(1))), "__iter__");
                                while (true) {
                                    try {
                                        var cursorD = (PyObject) e.invokeMethod(cursorDIt, "__next__");
                                        cursor.setItem(projAxis, cursorD);
                                        var slice = (PyObject) e.eval("slice");
                                        var someList = cursor.getItem(slice.call(e.getNone(), e.getNone(), e.fromJava(-1)));
                                        volume.setItem(e.newPyTuple(someList.getItem(e.fromJava(0)), someList.getItem(e.fromJava(1)), someList.getItem(e.fromJava(2))), maskColor);
                                    } catch (Exception ex) {
                                        break;
                                    }
                                }
                            } catch (Exception ex) {
                                break;
                            }
                        }
                    } catch (Exception ex) {
                        break;
                    }
                }
            } catch (Exception ex) {
                break;
            }
        }

        //Perform iterative shape carving until convergence
        var voxelsRemoved = e.fromJava(1);
        var cond = (PyObject) (e.invokeMethod(voxelsRemoved, "__gt__", e.fromJava(0)));
        while (cond.isTrue()) {
            voxelsRemoved = e.fromJava(0);
            var dItNew = e.invokeMethod(e.invokeFunction("range", e.fromJava(3)), "__iter__");
            while (true) {
                try {
                    var projAxis = (PyObject) e.invokeMethod(dItNew, "__next__");
                    var u = (PyObject) e.invokeMethod((PyObject) e.invokeMethod(projAxis, "__add__", e.fromJava(1)), "__mod__", e.fromJava(3)); // other axis 0
                    var v = (PyObject) e.invokeMethod((PyObject) e.invokeMethod(projAxis, "__add__", e.fromJava(2)), "__mod__", e.fromJava(3)); // other axis 1

                    //Do front/back sweep
                    var sideIdxIt = e.invokeMethod(e.invokeFunction("range", e.fromJava(2)), "__iter__");
                    while (true) {
                        try {
                            var sideIdx = (PyObject) e.invokeMethod(sideIdxIt, "__next__");
                            var sweepDirection = e.invokeMethod(e.invokeMethod(sideIdx, "__mul__", e.fromJava(-2)), "__add__", e.fromJava(1));
                            var idxTuple = e.newPyTuple(projAxis, sideIdx);
                            var skipThis = skip.getItem(idxTuple);
                            if (skipThis.isTrue()) {
                                continue;
                            }

                            var view = views.getItem(idxTuple);
                            var depth = depths.getItem(idxTuple);

                            var cursorVIt = e.invokeMethod(e.invokeFunction("range", volumeDims.getItem(v)), "__iter__");
                            while (true) {
                                try {
                                    var cursorV = (PyObject) e.invokeMethod(cursorVIt, "__next__");
                                    cursor.setItem(v, cursorV);
                                    var cursorUIt = e.invokeMethod(e.invokeFunction("range", volumeDims.getItem(u)), "__iter__");
                                    while (true) {
                                        try {
                                            var cursorU = (PyObject) e.invokeMethod(cursorUIt, "__next__");
                                            cursor.setItem(u, cursorU);
                                            //March along ray
                                            var start = depth.getItem(e.newPyTuple(cursor.getItem(v), cursor.getItem(u)));
                                            var isForwardSweep = (PyObject) e.invokeMethod(sweepDirection, "__eq__", e.fromJava(1));
                                            var stop = isForwardSweep.isTrue() ? volumeDims.getItem(projAxis) : e.fromJava(-1);
                                            var cursorDIt = e.invokeMethod(e.invokeFunction("range", start, stop, sweepDirection), "__iter__");
                                            while (true) {
                                                try {
                                                    var cursorD = (PyObject) e.invokeMethod(cursorDIt, "__next__");
                                                    cursor.setItem(projAxis, cursorD);
                                                    //Read volume color
                                                    var volIdx = e.newPyTuple(cursor.getItem(e.fromJava(2)), cursor.getItem(e.fromJava(1)), cursor.getItem(e.fromJava(0)));
                                                    var color = volume.getItem(volIdx);
                                                    var isMaskColor = (PyObject) e.invokeMethod(color, "__eq__", maskColor);
                                                    if (isMaskColor.isTrue()) {
                                                        continue;
                                                    }
                                                    volume.setItem(volIdx, view.getItem(e.newPyTuple(cursor.getItem(v), cursor.getItem(u))));
                                                    color = volume.getItem(volIdx);

                                                    //Check photo-consistency of volume at cursor
                                                    var isConsistent = e.getTrue();
                                                    var checkAxisIt = e.invokeMethod(e.invokeFunction("range", e.fromJava(3)), "__iter__");
                                                    while (true) {
                                                        try {
                                                            var checkAxis = (PyObject) e.invokeMethod(checkAxisIt, "__next__");
                                                            var checkU = (PyObject) e.invokeMethod(e.invokeMethod(checkAxis, "__add__", e.fromJava(1)), "__mod__", e.fromJava(3));
                                                            var checkV = (PyObject) e.invokeMethod(e.invokeMethod(checkAxis, "__add__", e.fromJava(2)), "__mod__", e.fromJava(3));
                                                            var checkSideIt = e.invokeMethod(e.invokeFunction("range", e.fromJava(2)), "__iter__");
                                                            while (true) {
                                                                try {
                                                                    var checkSide = (PyObject) e.invokeMethod(checkSideIt, "__next__");
                                                                    idxTuple = e.newPyTuple(checkAxis, checkSide);
                                                                    var skipNewThis = skip.getItem(idxTuple);
                                                                    if (skipNewThis.isTrue()) {
                                                                        continue;
                                                                    }
                                                                    var idxTupleInner = e.newPyTuple(checkAxis, checkSide, cursor.getItem(checkV), cursor.getItem(checkU));
                                                                    var checkViewColor = views.getItem(idxTupleInner);
                                                                    var checkViewDepth = depths.getItem(idxTupleInner);
                                                                    var isMaxSide = (PyObject) e.invokeMethod(checkSide, "__ne__", e.fromJava(0));
                                                                    var isDepthOccluded = isMaxSide.isTrue() ? (PyObject) e.invokeMethod(checkViewDepth, "__le__", cursor.getItem(checkAxis)) : (PyObject) e.invokeMethod(cursor.getItem(checkAxis), "__le__", checkViewDepth);
                                                                    if (isDepthOccluded.isTrue()) {
                                                                        var isColorDifferent = (PyObject) e.invokeMethod(checkViewColor, "__ne__", color);
                                                                        if (isColorDifferent.isTrue()) {
                                                                            isConsistent = e.getFalse();
                                                                            break;
                                                                        }
                                                                    }
                                                                } catch (Exception ex) {
                                                                    break;
                                                                }
                                                            }
                                                            if (isConsistent.isFalse()) {
                                                                break;
                                                            }
                                                        } catch (Exception ex) {
                                                            break;
                                                        }
                                                    }
                                                    if (isConsistent.isTrue()) {
                                                        break;
                                                    }

                                                    //Clear out voxel
                                                    voxelsRemoved = (PyObject) e.invokeMethod(voxelsRemoved, "__add__", e.fromJava(1));
                                                    volume.setItem(volIdx, maskColor);
                                                } catch (Exception ex) {
                                                    break;
                                                }
                                            }
                                            //Update depth value
                                            depth.setItem(e.newPyTuple(cursor.getItem(v), cursor.getItem(u)), cursor.getItem(projAxis));
                                        } catch (Exception ex) {
                                            break;
                                        }
                                    }
                                } catch (Exception ex) {
                                    break;
                                }
                            }
                        } catch (Exception ex) {
                            break;
                        }
                    }
                } catch (Exception ex) {
                    break;
                }
            }
            cond = (PyObject) (e.invokeMethod(voxelsRemoved, "__gt__", e.fromJava(0)));
        }

        ////Do a final pass to fill in any missing colors
        volume.setItem((PyObject) e.invokeMethod(volume, "__lt__", e.fromJava(0)), e.fromJava(0xff00ff));

        return volume;
    }

    // views has shape like (6, 16 * 16), but for NumPy it gets (3, 2, 16, 16)
    public static PyObject viewsToNumPy(PythonScriptEngine e, List<List<Integer>> views) throws ScriptException, NoSuchMethodException {
        var pyObjArr2D = new PyObject[6];
        for (var i = 0; i < 6; i++) {
            var pyObjArr1D = new PyObject[views.getFirst().size()];
            for (var j = 0; j < views.getFirst().size(); j++) {
                pyObjArr1D[j] = (PyObject) e.invokeFunction("int", views.get(i).get(j));
            }
            var pyList1D = e.newPyList(pyObjArr1D);
            pyObjArr2D[i] = pyList1D;
        }
        var pyList2D = e.newPyList(pyObjArr2D);
        e.put("views", pyList2D);
        var sideLen = (int) Math.sqrt(views.getFirst().size());
        return (PyObject) e.invokeMethod(e.eval("np.array(views)"), "reshape", e.fromJava(3), e.fromJava(2), sideLen, sideLen);
    }
}

File: src/main/java/org/example/shapecarvejavafx/ShapeCarver.kt
package org.example.shapecarvejavafx

import javafx.beans.property.SimpleIntegerProperty
import javafx.embed.swing.SwingFXUtils
import javafx.scene.image.*
import kotlinx.coroutines.channels.Channel
import java.io.File
import javax.imageio.ImageIO


class ShapeCarver(private var output: Output) {
    private var depths: MutableList<MutableList<Int>> = mutableListOf()
    private var x: IntArray = IntArray(3) // cursor

    val channel = Channel<Unit>()

    suspend fun carve(
        views: List<List<Int>>,  // 2d images
        maskColor: Int, skip: BooleanArray // views to skip
    ) {
        val volume = output.volume
        val dims = output.dims
        require(skip.isNotEmpty())

        // Initialize depth fields
        for (d in 0..2) {
            val u = (d + 1) % 3
            val v = (d + 2) % 3
            var s = 0
            while (s <= dims[d] - 1) {
                val vals = IntArray(dims[u] * dims[v])
                val view = views[depths.size]
                val sOp = if (s == 0) dims[d] - 1 else 0
                vals.forEachIndexed { i, _ ->
                    vals[i] = if (!skip[depths.size] && view[i] == maskColor) sOp else s
                }
                depths.add(vals.toMutableList())
                s += dims[d] - 1
            }

            // Clear out volume
            x[v] = 0
            while (x[v] < dims[v]) {
                x[u] = 0
                while (x[u] < dims[u]) {
                    x[d] = depths[2 * d + 1][x[u] + x[v] * dims[u]]
                    while (x[d] <= depths[2 * d][x[u] + x[v] * dims[u]]) {
                        volume[x[0] + dims[0] * (x[1] + dims[1] * x[2])].set(maskColor)
                        ++x[d]
                    }
                    ++x[u]
                }
                ++x[v]
            }
            channel.send(Unit)
        }

        // Perform iterative seam carving until convergence
        var removed = 1
        while (removed > 0) {
            removed = 0
            for (d in 0..2) {
                val u = (d + 1) % 3
                val v = (d + 2) % 3

                // Do front/back sweep
                for (s in listOf(-1, 1)) {
                    val vNum = 2 * d + if (s < 0) 1 else 0
                    if (skip[vNum]) {
                        continue
                    }

                    val view = views[vNum]
                    val depth = depths[vNum]

                    x[v] = 0
                    while (x[v] < dims[v]) {
                        x[u] = 0
                        while (x[u] < dims[u]) {
                            // March along ray
                            val bufIdx = x[u] + x[v] * dims[u]
                            x[d] = depth[bufIdx]
                            while (x[d] in 0 until dims[d]) {
                                // Read volume color
                                val volIdx = x[0] + dims[0] * (x[1] + dims[1] * x[2])
                                var color = volume[volIdx]
                                if (color.get() == maskColor) {
                                    x[d] += s
                                    continue
                                }

                                volume[volIdx].set(view[x[u] + dims[u] * x[v]])
                                color = volume[volIdx]

                                // Check photo-consistency of volume at x
                                var consistent = true
                                for (a in 0..2) {
                                    val b = (a + 1) % 3
                                    val c = (a + 2) % 3
                                    val idx = x[b] + dims[b] * x[c]
                                    for (t in 0..1) {
                                        val fnum = 2 * a + t
                                        if (skip[fnum]) continue
                                        val fcolor = views[fnum][idx]
                                        val fdepth = depths[fnum][idx]
                                        if (if (t != 0) fdepth <= x[a] else x[a] <= fdepth) {
                                            if (fcolor != color.get()) {
                                                consistent = false
                                                break
                                            }
                                        }
                                    }
                                    if (!consistent) break
                                }
                                if (consistent) break

                                // Clear out voxel
                                ++removed
                                volume[volIdx].set(maskColor)
                                x[d] += s
                            }

                            // Update depth value
                            depth[bufIdx] = x[d]
                            ++x[u]
                        }
                        ++x[v]
                    }
                }
                channel.send(Unit)
            }
        }

        // Do a final pass to fill in any missing colors
        var n = 0 // linear index. See loop invariant below
        x[2] = 0
        while (x[2] < dims[2]) {
            x[1] = 0
            while (x[1] < dims[1]) {
                x[0] = 0
                while (x[0] < dims[0]) {
                    if (volume[n].get() < 0) {
                        volume[n].set(0xff00ff)
                    }
                    ++x[0]
                    ++n
                }
                ++x[1]
            }
            ++x[2]
        }

        val projections = output.project()
        output.saveProjectionsToImages(projections)
        channel.close()
    }
}

data class Output(val volume: List<SimpleIntegerProperty>, val dims: IntArray) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Output

        if (volume != other.volume) return false
        if (!dims.contentEquals(other.dims)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = volume.hashCode()
        result = 31 * result + dims.contentHashCode()
        return result
    }

    // project returns the six axis-oriented orthogonal views of the shape
    fun project(): List<List<List<Int>>> {
        val views = mutableListOf<List<List<Int>>>()
        for (d in 0..2) {
            val u = (d + 1) % 3
            val v = (d + 2) % 3
            val view = mutableListOf<List<Int>>()
            for (s in 0 until dims[d]) {
                val vals = mutableListOf<Int>()
                for (vIdx in 0 until dims[v]) {
                    for (uIdx in 0 until dims[u]) {
                        vals.add(volume[uIdx + dims[u] * (vIdx + dims[v] * s)].get())
                    }
                }
                view.add(vals)
            }
            views.add(view)
        }
        return views
    }

    fun saveProjectionsToImages(views: List<List<List<Int>>>) {
        for ((i, view) in views.withIndex()) {
            val img = projectionToImage(view)
            val legacyImg = SwingFXUtils.fromFXImage(img, null)
            ImageIO.write(legacyImg, "png", File("view$i.png"))
        }

    }

    // projectionToImage takes a matrix of integers, where the integers represent 24-bit RGB colors without transparency
    private fun projectionToImage(view: List<List<Int>>): Image {
        val width = view[0].size
        val height = view.size
        val pixels = IntArray(width * height)
        for (y in 0 until height) {
            for (x in 0 until width) {
                val color = view[y][x]
                pixels[x + y * width] = color
            }
        }
        val argbPixels = IntArray(width * height)
        for (i in pixels.indices) {
            argbPixels[i] = 0xff shl 24 or pixels[i]
        }

        val pixelData = ByteArray(width * height * 4) // Assuming 4 bytes per pixel (ARGB)


        // Fill the pixelData with data from argbPixels
        for (y in 0 until height) {
            for (x in 0 until width) {
                val index = (y * width + x) * 4
                val argb = argbPixels[y * width + x]
                // first comes A
                pixelData[index] = (argb shr 24).toByte()
                // then R
                pixelData[index + 1] = (argb shr 16).toByte()
                // then G
                pixelData[index + 2] = (argb shr 8).toByte()
                // and lastly B
                pixelData[index + 3] = argb.toByte()
            }
        }

        val image = WritableImage(width, height)
        val pixelWriter = image.pixelWriter
        pixelWriter.setPixels(0, 0, width, height, PixelFormat.getByteBgraInstance(), pixelData, 0, width * 4)
        return image
    }
}
